### ISSUE TITLE
Small Bug in GenerateSignatureBase causes failure in LTI Conformance Tests

### ISSUE THREAD

**Issue by wkfry** [LtiLibrary/LtiLibrary#14]
Hey Andy,

First of all, thanks for your excellent work on LTI and support for .NET. Great stuff.

I believe that I found a small bug in your implementation for generating your OAuth signature. I believe that IMS recently changed their LTI Test Suite and it now exposes the (uncommon) scenario that exposes the bug. If there are Query String parameters, they must be included in the NameValueCollection when generating the Signature.

This was a quick fix that worked and allowed me to pass the test suite. I think it would have been better for me to clone the NameValueCollection, but since I don't think this will happen in the real world often, I didn't bother.

Just thought I'd let you know! Once again, thanks!

```
    public static string GenerateSignatureBase(string httpMethod, Uri url, NameValueCollection parameters)
    {
        // RFC 5849 3.4.1.1
        var signatureBase = new StringBuilder();
        signatureBase.Append(httpMethod.ToRfc3986EncodedString().ToUpperInvariant()).Append('&');

        // RFC 5849 3.4.1.2
        var normalizedUrl = string.Format("{0}://{1}", url.Scheme.ToLowerInvariant(), url.Host.ToLowerInvariant());
        if (!((url.Scheme == "http" && url.Port == 80) || (url.Scheme == "https" && url.Port == 443)))
        {
            normalizedUrl += ":" + url.Port;
        }
        normalizedUrl += url.AbsolutePath;
        signatureBase.Append(normalizedUrl.ToRfc3986EncodedString()).Append('&');

        // Add URL Parameters
        if (!string.IsNullOrEmpty(url.Query))
        {
            // According to the OAuth 1.0 specification, all Query String parameters should be
            // added into the set of the name/value pairs so they are included in the
            // signature.
            //
            // This is a rare, edge case; however, the latest LTI Certification actually tests
            // for this. The following code fixes this problem.

            string[] queryParts = url.Query.Substring(1).Split('&');

            foreach (string part in queryParts)
            {
                string name;
                string value;

                int indexOfEquals = part.IndexOf('=');

                if (indexOfEquals == -1)
                {
                    name = part;
                    value = "";
                }
                else
                {
                    name = part.Substring(0, indexOfEquals);
                    value = part.Substring(indexOfEquals + 1);
                }

                name = Uri.UnescapeDataString(name);
                value = Uri.UnescapeDataString(value);

                parameters.AddParameter(name, value);
            }
        }

        // Per RFC 5849 3.4.1.3, do not include the OAuth signature or realm in the signature base string
        var excludedNames = new List<string> {OAuthConstants.SignatureParameter, OAuthConstants.RealmParameter};

        // Construct the signature string
        signatureBase.Append(parameters.ToNormalizedString(excludedNames).ToRfc3986EncodedString());

        return signatureBase.ToString();
    }
```

Bill

**Comment by srijken** [comment]
@wkfry Can you verify this test case? Is the url being asserted indeed correct (it is the result of the above code)

```
        [Fact]
        public void GenerateSignatureBase_WithParametersAndQueryString()
        {
            var parameters = new NameValueCollection();
            parameters["key1"] = "value1";
            parameters["key2"] = "value2";

            string result = OAuthUtility.GenerateSignatureBase("POST", new Uri("http://example.com/test/1?id=5"), parameters);

            Assert.Equal("POST&http%3A%2F%2Fexample.com%2Ftest%2F1&id%3D5%26key1%3Dvalue1%26key2%3Dvalue2", result);
        }
```

**Comment by andyfmiller** [comment]
Interesting development. I had to explicitly remove them in the past to pass the test...which I always thought was weird. I even brought that up at an IMS meeting years ago and was told we should not be passing querystring parameters in the LTI URL.

I'll verify and make the change permanent (or let you know if I discover something different).

**Comment by andyfmiller** [comment]
@wkfry I just tried your code, but the first certification test failed because the OAuth signature was invalid. Then I removed your code, and the test passed.

Here is a screenshot (ignore 2 and 3...the certification server cannot reach my local machine).

![oauthsignature](https://cloud.githubusercontent.com/assets/2666882/13590285/5e03f918-e493-11e5-9c8c-43aee0f03182.png)

**Comment by wkfry** [comment]
That's weird. I had been passing the test suite without a problem and then one day, it started to fail until I added this code. I wonder if they had a temporary problem with the test suite. I'm very sorry for the confusion.

**Comment by spvickers** [comment]
Andy, if your code is excluding any query parameters on a launch URL from the OAuth signing process, then this should be changed.  All parameters (GET and POST) must be included when signing a message as per the OAuth spec.

**Comment by andyfmiller** [comment]
Hello Stephen,

The querystring parameters are being added to the OAuth signing process,

https://github.com/andyfmiller/LtiLibrary/blob/master/LtiLibrary.Core/OAuth/OAuthRequest.cs#L211

My memory was faulty when I wrote my comment above. I don't know what happened to wkfry or why his code worked for him. It would have resulted in the params being added twice unless Parameters eliminates duplicates (which it may...I can't remember offhand).

Thanks for checking in. http://lti.tools was invaluable over the last few days shaking bugs out of my Outcomes-2 implementation. I could not track down the latest for 3 days before I finally noticed I forgot the "p1" in the LISResult media type.

**Comment by jeroenhabets** [comment]
Hello Stephen (@spvickers), Andy (@andyfmiller),

Yesterday we actually hosted a meeting between two LMS vendors and one Exam (LTI) provider about this very topic.

For the LTI launch the spec is unambiguous: add query string parameters during OAuth signing.
However, for the Outcomes 1.0 callback unfortunately the spec is **not** that clear and unequivocal (OAuth bodyhash): 

[LTI Implementation Guide v1.14.- 3 Security for application/xml Messages](https://www.imsglobal.org/specs/ltiv1p1/implementation-guide#toc-16)
"All of the OAuth parameters, HTTP method, and URL are signed like any other OAuth signed request.  Other than in producing the body hash value, the actual POST data is not involved in the computation of the oauth_signature." 

Leading to two questions to you both:
1) When posting the Outcomes 1.0 message should the URL parameters be included during OAuth signing or not? (I've been informed Andy's library released version does not, as attested by the creation of [this issue #14](https://github.com/andyfmiller/LtiLibrary/issues/14))

2) If so (the answer I would expect), and the LTI certification test does not correctly handle it, it should be fixed. if it currently fails for consumers including the parameters in the OAuth signature [Wrong: should succeed], and succeeds for consumers failing to include them [Wrong: should fail]
Andy's last comment makes it unclear for me, what is currently implemented in the certification  suite.

Plus one question to @andyfmiller if your implementation was indeed incorrect when will a fix for both Outcomes 1.0 and 2.0 become available?

NOTE: Our question is for Outcomes 1.0 but we also support Outcomes 2.0 so I assume the same applies there.

Kind regards,

Jeroen Habets

**Comment by spvickers** [comment]
With regard to the signing of any LTI message (or requests), the LTI specification has currently adopted the OAuth 1 specification, including the revision for body hashing messages which are not of an application/x-www-form-urlencoded type.  In all cases the signature is based on the complete URL of the endpoint (including the query string), the HTTP action (e.g. POST) and the body (either as POST data parameters or a hash of the body if XML or JSON, for example).  I am not aware of any problem with the IMS LTI certification suite failing to follow this procedure - please let me have more details if you have experienced an issue here.  Thanks.

**Comment by MountyMartin** [comment]
See https://github.com/andyfmiller/LtiLibrary/blob/master/LtiLibrary.Core/Outcomes/v1/OutcomesClient.cs#L296 

It doesn't add the querystring parameters to the OAuth signing process, it misses the 

`var querystring = new UrlEncodingParser(Uri.UnescapeDataString(Url.Query));
            parameters.Add(querystring);`

The outcomes client uses a different code path to get to the OAuthUtility.GenerateSignature than the one Andy mentions in https://github.com/andyfmiller/LtiLibrary/blob/master/LtiLibrary.Core/OAuth/OAuthRequest.cs#L211

**Comment by spvickers** [comment]
Thanks MountyMartin, I suspect your comment is intended for Andy's benefit.  I am not aware that this code has been certified by IMS so assume it does not represent an issue with the IMS certification process.  But please let me know if I have this wrong.

**Comment by jeroenhabets** [comment]
@spvickers thank you for your clarification. After my post I also found your answer at the IMS forum (https://www.imsglobal.org/forum/learning-tools-interoperability-public-forum/137206)

Because @andyfmiller had written that he had to omit the request parameters from the hash in order to succesfully test against the IMS Certification Suite, I am still not certain if the suite does this correctly but for now let's wait on @andyfmiller's take  on this.

**Comment by MountyMartin** [comment]
@spvickers  The IMS forum post was from me. I wanted to reply on the forum but it gives me an error "We're sorry, but you can't post comments here!". Weird because I was logged in. Anyway, thanks for the reply.

Regards,

Martijn van den Berg

**Comment by andyfmiller** [comment]
I'll re-open this issue to make sure I test it thoroughly.

**Comment by andyfmiller** [comment]
Thank you @MountyMartin for finding the problem...it will not surprise you that I could reproduce it at will. I just checked in a fix which actually simplified the code (I love it when that happens). Initial testing looks good and it will be included in the next NuGet package.

I won't actually close this issue until the whole package is tested thoroughly.

**Comment by andyfmiller** [comment]
Fixed in 1.6 (now available on NuGet).
